apiserver/facades/agent/uniter/uniter.go

```diff
@@ -692,12 +692,21 @@ func (u *UniterAPI) CharmURL(args params.Entities) (params.StringBoolResults, er
 			var unitOrApplication state.Entity
 			unitOrApplication, err = u.st.FindEntity(tag)
 			if err == nil {
-				charmURLer := unitOrApplication.(interface {
-					CharmURL() (*charm.URL, bool)
-				})
-				curl, ok := charmURLer.CharmURL()
-				if curl != nil {
-					result.Results[i].Result = curl.String()
+				var cURL *charm.URL
+				var ok bool
+
+				switch entity := unitOrApplication.(type) {
+				case *state.Application:
+					cURL, ok = entity.CharmURL()
```

Comment:
Weird, looks like we need more work to make this a common interface

Reply:
Yes, this only does the work required for the unit case, but we should address the other docs that store a `*charm.URL`, at which time a lot of double handling can be eliminated.

Reply:
The name of `ok` is misleading here. For the unit, it was referring to the charm url pointer, which has been appropriately changed to an error. For application, the bool is whether a charm has been forced or not.

This facade is shared between the unit and application part of the uniter api - on the unit side, the bool is ignored. For the uniter application, force is used by the RemoteStateWorker.